File: Demo/Rxjava/RxJavaJavaDemo/app/src/main/java/com/example/rxjavademo/MainActivity.java
package com.example.rxjavademo;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.BiFunction;
import io.reactivex.functions.Function;
import io.reactivex.functions.Predicate;
import io.reactivex.schedulers.Schedulers;
import model.Base;
import model.Card;
import model.City;
import model.Mall;
import util.Factory;
import util.FactoryCity;
import util.FactoryMall;

public class MainActivity extends AppCompatActivity {
    private final static String TAG = "MainActivity";

    private final static String[] CITY = {"深圳市", "广州市", "珠海市"};
    private final static String[] MALL = {"麦当劳", "星巴克", "真功夫", "肯德基"};

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Button btn = findViewById(R.id.HiButton);
        btn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                final List<Base> cityList = createBaseList(CITY, new FactoryCity());

                final List<Base> mallList = createBaseList(MALL, new FactoryMall());

                Observable<Base> city = createObservable(cityList);

                Observable<Base> mall = createObservable(mallList);

                Observer<String> card = createObserver();

                mSubscribe(city, mall, card);
            }
        });
    }

    private List<Base> createBaseList(String[] mData, Factory mFactory) {
        final List<Base> result = new ArrayList<>();
        for(String s : mData){
            result.add(mFactory.create(s));//这里数据类型会 向上转型为 Base
        }
        return result;
    }

    private Observable<Base> createObservable(final List<Base> baseList) {
        return Observable.create(new ObservableOnSubscribe<Base>() {
            @Override
            public void subscribe(ObservableEmitter<Base> emitter) {
                for (Base c : baseList) {
                    Log.d(TAG, mPrint(c.getName(), " 发送事件："));
                    emitter.onNext(c);
                }
            }
        }).subscribeOn(Schedulers.newThread());
    }

    private Observer<String> createObserver() {
        return new Observer<String>() {//观察者
            @Override
            public void onSubscribe(Disposable d) {
                Log.d(TAG, "onSubscribe");
            }

            @Override
            public void onNext(String s) {//响应事件
                Log.d(TAG, mPrint(s, " 响应事件："));
            }

            @Override
            public void onError(Throwable e) {
                Log.d(TAG, "onError");
            }

            @Override
            public void onComplete() {
                Log.d(TAG, "onComplete");
            }
        };
    }

    private void mSubscribe(Observable<Base> cityObservable, Observable<Base> mallObservable, Observer<String> mObserver) {

        //使用zip操作符将事件1：1合并
        Observable.zip(cityObservable, mallObservable, new BiFunction<Base, Base, Card>() {
            @Override
            public Card apply(Base city, Base mall) throws Exception {
                return new Card(city.getName(), mall.getName());
            }
        }).observeOn(AndroidSchedulers.mainThread()//观察者在主线程输出
        ).filter(new Predicate<Card>() {//过滤 深圳市
            @Override
            public boolean test(Card card) {
                return card.getCityName().equals("深圳市");
            }
        }).map(new Function<Card, String>() {//把卡片转换为字符串
            @Override
            public String apply(Card card) {
                return card.toString();
            }
        }).subscribe(mObserver);

    }

    private String mPrint(String msg, String type) {
        return "线程id：" + Thread.currentThread().getId() + type + msg;
    }
}

File: Demo/Rxjava/RxJavaKotlinDemo/app/src/main/java/com/example/rxjavademo/MainActivity.kt
package com.example.rxjavademo

import android.os.Bundle
import android.util.Log
import android.widget.Button
import androidx.appcompat.app.AppCompatActivity
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.functions.BiFunction
import io.reactivex.schedulers.Schedulers
import kotlinx.coroutines.*
import model.Card
import model.City
import model.Mall
import java.util.*

import util.RxUtil

class MainActivity : AppCompatActivity(), CoroutineScope by MainScope() { //协程结构化并发

    companion object {
        private const val TAG = "MainActivity"
        private val CITY = arrayOf("深圳市", "广州市", "珠海市")
        private val MALL = arrayOf("麦当劳", "星巴克", "真功夫", "肯德基")
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val btn = findViewById<Button>(R.id.HiButton)

        btn.setOnClickListener {
            launch {
                /* 当用户手抖狂点 button 时会发生什么？防止抖动？
                 * Observable.create<T> 在多个线程/协程中被调用可能会导致什么多线程问题？
                 */

                val cityAsync = async { initCity() }
                val mallAsync = async { initMall() }

                /* 经试验
                 *   await() 调用会阻塞等待 async 执行完毕获得返回值后，再继续执行
                 *   所以不能连接调用 async{}.await()，因为这会使 async 毫无作用
                 */
                val mallObservable = mallAsync.await()
                val cityObservable = cityAsync.await()

                val card = RxUtil().createObserver()

                //把 subscribe 提取出来使代码可读性更高
                mObservable(cityObservable, mallObservable).subscribe(card)
            }
        }
    }

    private suspend fun initCity(): Observable<City> {
        //这是高级函数，区分 sam 接口转化，1.3 及之前版本也可以这样写
        val cityList = createCityList { City(it) }
        return RxUtil().createObservable(cityList)//泛型
    }

    private suspend fun initMall(): Observable<Mall> {
        val mallList = createMallList { Mall(it) }
        return RxUtil().createObservable(mallList)
    }

    private suspend fun createCityList(createCity: (s: String) -> City): List<City> {
        val result: MutableList<City> = ArrayList()
        for (s in CITY) {
            RxUtil().mPrint("创建 $s", "Create City")
            delay(1000)
            result.add(createCity(s))
        }
        return result
    }

    private suspend fun createMallList(createMall: (s: String) -> Mall): List<Mall> {
        val result: MutableList<Mall> = ArrayList()
        for (s in MALL) {
            RxUtil().mPrint("创建 $s", "Create Mall")
            delay(1000)
            result.add(createMall(s))
        }
        return result
    }

    private fun createObserver(): Observer<String> {
        return object : Observer<String> {
            //观察者
            override fun onSubscribe(d: Disposable) {
                Log.d(TAG, "onSubscribe")
            }

            override fun onNext(s: String) { //响应事件
                mPrint(s, " 响应事件")
            }

            override fun onError(e: Throwable) {
                Log.d(TAG, "onError")
            }

            override fun onComplete() {
                Log.d(TAG, "onComplete")
            }
        }
    }

    private fun <T> createObservable(baseList: List<T>): Observable<T> {
        return Observable.create<T> { //emitter ->
            for (c in baseList) {
                mPrint(c.toString(), " 发送事件")
                it.onNext(c)
            }
        }.subscribeOn(Schedulers.newThread())
    }


    private fun <C, M> mObservable(cityObservable: Observable<C>, mallObservable: Observable<M>): Observable<String> {
        return Observable.zip(cityObservable, mallObservable, BiFunction<C, M, Card> { city, mall ->
            Card(city.toString(), mall.toString()) }
        ).observeOn(AndroidSchedulers.mainThread() //观察者在主线程输出
        ).filter { card ->
            card.cityName == "深圳市"
        }.map { card ->
            card.toString()
        }
    }

    //表达式返回，字符串模板
    private fun mPrint(msg: String, type: String): Int =
            Log.d(TAG,"线程id：${Thread.currentThread().id} $type: $msg")

    override fun onDestroy() {
        super.onDestroy()
        cancel()//作用域取消
    }
}